### PR TITLE
formula_cellar_checks: check keg for mismatched arches

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

There have been a few instances I've noticed that we've been silently
installing binaries built for x86_64 on ARM. There's probably more that
I haven't found yet, so it seems useful to check this with an audit.

This is usually a sign that the formula is installing pre-built binaries.
I'm not sure if the error from the audit should reflect this, so I've left
it out for now. Happy to tweak the wording as needed.

### CODE REVIEW

File: Library/Homebrew/os/linux/elf.rb [hunk 1]
```diff
@@ -33,7 +33,7 @@ module ELFShim
   private_constant :ARCHITECTURE_POWERPC
   ARCHITECTURE_ARM = 0x28
   private_constant :ARCHITECTURE_ARM
-  ARCHITECTURE_X86_64 = 0x62
+  ARCHITECTURE_X86_64 = 0x3E
```
Comment (carlocab) [inline Library/Homebrew/os/linux/elf.rb:36]: This is needed because checking the `arch` of ELF files on Linux currently returns `:dunno`.

File: Library/Homebrew/formula_cellar_checks.rb [hunk 2]
```diff
@@ -314,6 +314,24 @@ def check_cpuid_instruction(formula)
...
   end
 
+  def check_binary_arches(formula)
+    return unless formula.prefix.directory?
+    # There is no `binary_executable_or_library_files` method for the generic OS
+    return if !OS.mac? && !OS.linux?
+
+    keg = Keg.new(formula.prefix)
+    mismatches = keg.binary_executable_or_library_files.reject do |file|
+      file.arch == Hardware::CPU.arch
```
Comment (carlocab) [inline Library/Homebrew/formula_cellar_checks.rb:324]: `Hardware::CPU.arch` returns `:x86_64` when running under Rosetta, so this audit should behave as expected on ARM in `/usr/local` too.

File: Library/Homebrew/formula_cellar_checks.rb [hunk 3]
```diff
@@ -314,6 +314,24 @@ def check_cpuid_instruction(formula)
     "No `cpuid` instruction detected. #{formula} should not use `ENV.runtime_cpu_detection`."
   end
 
+  def check_binary_arches(formula)
+    return unless formula.prefix.directory?
+    # There is no `binary_executable_or_library_files` method for the generic OS
+    return if !OS.mac? && !OS.linux?
```
Comment (Rylan12) [inline Library/Homebrew/formula_cellar_checks.rb:320]: I'm just curious, why is this the case? I thought the default for this was to use the Linux version and, on macOS, the method was overridden. Is that not what happens with `--generic`?

Comment (carlocab) [reply Library/Homebrew/formula_cellar_checks.rb:320]: Yes. The default behaviour is to call the method `elf_files`, but that's Linux-only too.

File: Library/Homebrew/formula_cellar_checks.rb (hunk 2 above)
Comment (carlocab) [reply Library/Homebrew/formula_cellar_checks.rb:324]: Though, one corner case worth considering, where I don't know what happens (and don't have the hardware to test...): a user who has an M1 machine with `brew` installed into `/usr/local`, and runs `/usr/local/bin/brew install` without running the process under Rosetta2. I suspect this audit will produce errors for them, but I don't know if `brew` will error out for that case earlier than that.

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_cellar_checks.rb:324]: > I don't know if `brew` will error out for that case earlier than that.

it will 🎉

**Review by MikeMcQuaid — APPROVED** [review]
Great work 🎉
